src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs

```diff
@@ -464,9 +475,87 @@ impl<'tcx> RegionInferenceContext<'tcx> {
             },
         }
 
+        self.add_static_impl_trait_suggestion(
+            infcx, &mut diag, fr_name, fr_region, outlived_fr_region
+        );
+
         diag.buffer(errors_buffer);
     }
 
+    fn add_static_impl_trait_suggestion(
+        &self,
+        infcx: &InferCtxt<'_, '_, 'tcx>,
+        diag: &mut DiagnosticBuilder<'_>,
+        fr_name: RegionName,
+        fr_region: Option<Region<'tcx>>,
+        outlived_fr_region: Option<Region<'tcx>>,
+    ) {
+        if let (Some(f), Some(ty::RegionKind::ReStatic)) = (fr_region, outlived_fr_region) {
+            if let Some(ty::TyS {
+                sty: ty::TyKind::Anon(did, substs),
+                ..
+            }) = infcx.tcx.is_suitable_region(f)
+                    .map(|r| r.def_id)
+                    .map(|id| infcx.tcx.return_type_impl_trait(id))
+                    .unwrap_or(None)
+            {
+                let has_static_predicate = {
+                    let predicates_of = infcx.tcx.predicates_of(*did);
+                    let bounds = predicates_of.instantiate(infcx.tcx, substs);
+
+                    let mut found = false;
+                    for predicate in bounds.predicates {
+                        if let ty::Predicate::TypeOutlives(binder) = predicate {
+                            if let ty::OutlivesPredicate(
+                                _,
+                                RegionKind::ReStatic
+                            ) = binder.skip_binder() {
+                                found = true;
+                                break;
+                            }
+                        }
+                    }
+
+                    found
+                };
+
+                debug!("add_static_impl_trait_suggestion: has_static_predicate={:?}",
+                       has_static_predicate);
+                let static_str = keywords::StaticLifetime.name();
+                if has_static_predicate {
+                    let span = self.get_span_of_named_region(infcx.tcx, f, &fr_name);
+                    if let Ok(snippet) = infcx.tcx.sess.source_map().span_to_snippet(span) {
+                        diag.span_suggestion(
+                            span,
+                            &format!(
+                                "you can add a constraint to the definition of `{}` to require it \
+                                 outlive `{}`",
```

Comment:
Is this always going to have the form `'a: 'static`? If so, I think I would rather that we suggest that they just replace uses of `'a` with `'static`. 

But for type parameters, `T: 'static` makes sense, of course.

Reply:
Fixed this in most recent commit.